src/providers/aws.rs

```diff
@@ -0,0 +1,858 @@
+                Region::EuCentral1 => "ami-0cdab515472ca0bac",   // Frankfurt
+                Region::EuNorth1 => "ami-c37bf0bd",              // Stockholm
+                Region::EuWest1 => "ami-01cca82393e531118",      // Ireland
+                Region::EuWest2 => "ami-0a7c91b6616d113b1",      // London
+                Region::EuWest3 => "ami-033e0056c336ecff0",      // Paris
+                Region::SaEast1 => "ami-094c359b4d8c6a8ca",      // Sao Paulo
+                Region::UsEast1 => "ami-064a0193585662d74",      // N Virginia
+                Region::UsEast2 => "ami-021b7b04f1ac696c2",      // Ohio
+                Region::UsWest1 => "ami-056d04da775d124d7",      // N California
+                Region::UsWest2 => "ami-09a3d8a7177216dcf",      // Oregon
+                x => panic!("Unsupported Region {:?}", x),
+            }
+            .into(),
+        )
+    }
+}
+
+impl Into<String> for UbuntuAmi {
+    fn into(self) -> String {
+        self.0
+    }
+}
+
+/// A descriptor for a particular machine setup in a tsunami.
+///
+/// The `setup` argument is called once for every spawned instances of this type with a handle
+/// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
+/// commands on the host in question.
+#[derive(Clone)]
+pub struct MachineSetup {
+    region: Region,
+    instance_type: String,
+    ami: String,
+    setup: Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl super::MachineSetup for MachineSetup {
+    type Region = String;
+
+    fn region(&self) -> Self::Region {
+        self.region.name().to_string()
+    }
+}
+
+impl PartialEq for MachineSetup {
+    fn eq(&self, other: &Self) -> bool {
+        self.ami == other.ami && self.instance_type == other.instance_type
+    }
+}
+
+impl Eq for MachineSetup {}
+
+impl std::hash::Hash for MachineSetup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.instance_type.hash(state);
+        self.ami.hash(state);
+    }
+}
+
+impl Default for MachineSetup {
+    fn default() -> Self {
+        MachineSetup {
+            region: Region::UsEast1,
+            instance_type: "t3.small".into(),
+            ami: UbuntuAmi::from(Region::UsEast1).into(),
+            setup: None,
+        }
+    }
+}
+
+impl MachineSetup {
+    /// Set up the machine in a specific EC2
+    /// [`Region`](http://rusoto.github.io/rusoto/rusoto_core/region/enum.Region.html).
+    ///
+    /// The default region is us-east-1. [Available regions are listed
+    /// here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions)
+    /// AMIs are region-specific. This will overwrite the ami field to
+    /// the Ubuntu 18.04 LTS AMI in the selected region.
+    pub fn region(mut self, region: Region) -> Self {
+        self.region = region.clone();
+        self.ami = UbuntuAmi::from(region).into();
```

Comment:
It's weird to me for the AMI to change just because I changed region. I guess I see why that'd be the case, but I think it'd be better to simply require that callers provide an AMI when they change regions.

Reply:
This avoids an annoying user problem when they change the region and then get some error about needing to call `ami()`. In most cases, the Ubuntu 18.04 ami is what people want and the region is what they want to change - better to optimize for the common case.

Reply:
I have never once wanted a stock AMI on EC2 for tsunami experiments, but maybe I'm the odd one out here? I'd personally prefer for us to not have to maintain a list of up-to-date AMIs though. And issuing a breaking change if they ever change. It seems more reasonable to instead just have `region` take a mandatory `ami` argument, no?

Reply:
it's better to have _some_ default (which is documented) than make everyone look up the ami id they want  (maintaining their own version of `UbuntuAmi`) and pass it in?

The only kind of confusing behavior here is something like
```
MachineSetup::default().ami("foobar").region("us-west-2")
```
where the ami gets overwritten. This is unfortunate but documented, and for use cases like the one you said where you are passing an ami id anyway, doesn't get in the way if you are ok with the default region.